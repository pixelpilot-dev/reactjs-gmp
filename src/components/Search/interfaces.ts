export interface ISearchProps {
  className?: string;
}
