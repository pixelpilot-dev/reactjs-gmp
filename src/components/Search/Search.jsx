import React, { PureComponent } from 'react';

import styles from './Search.module.scss';

export class Search extends PureComponent {
  state = {
    text: '',
  }

	handleChangeText = ({ target: { value } }) => {
		const { onChangeSearch } = this.props;

		this.setState({
			text: value,
		});

		onChangeSearch(value);
	}

	render() {
		const { text } = this.state;

		return (
			<input
				type="text"
				name="search"
				className={styles.input}
				placeholder="What do you want to watch?"
				value={text}
				onChange={this.handleChangeText}
			/>
		);
	}
};
