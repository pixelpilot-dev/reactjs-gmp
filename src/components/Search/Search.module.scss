.input {
  width: 100%;
  height: 57px;
  background: rgba(50, 50, 50, 0.8);
  opacity: 0.7;
  border-radius: 4px;
  border: 0;
  font: {
    family: 'Montserrat', sans-serif;
    weight: 500;
    size: 20px;
  };
  line-height: 24px;
  color: #fff;
  opacity: 0.3;
  margin-bottom: 10px;
  padding: 0 19px;

  &::placeholder {
    color: #fff;
    opacity: 0.3;
  }
}