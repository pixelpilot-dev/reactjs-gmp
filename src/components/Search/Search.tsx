import React, { useEffect, useState } from 'react';
import cn from 'classnames';
import { useNavigate, useParams } from 'react-router-dom';
import { Formik, Form } from 'formik';
import { I18Y, LOCALE } from '../../core/i18y';
import { ISearchProps } from './interfaces';
import { Button } from '../UI/Button';
import { Input } from '../UI/Input';

import styles from './Search.module.scss';

export const Search: React.FC<ISearchProps> = ({ className }) => {
  const navigate = useNavigate();
  const { searchQuery } = useParams();
  const [search, setSearch] = useState('');

  useEffect(() => {
    if (searchQuery) {
      setSearch(searchQuery);
    }
  }, [searchQuery]);

  const onSubmit = (values, setErrors) => {
    navigate(`/search/${values.search}`, { replace: true });
  };

  return (
    <div className={cn(styles.search, className)}>
      <div className={styles.title}>{I18Y[LOCALE].SEARCH_TITLE}</div>

      <Formik
        enableReinitialize
        initialValues={{ search }}
        onSubmit={(values, { setErrors }) => {
          onSubmit(values, setErrors);
        }}
      >
        {props => {
          const { values } = props;

          return (
            <Form className={styles.form}>
              <Input
                type='text'
                name='search'
                value={values.search}
                placeholder={I18Y[LOCALE].SEARCH_PLACEHOLDER}
                id='search'
                className={styles.input}
              />

              <Button type='submit' theme='default' className={styles.button}>
                {I18Y[LOCALE].SEARCH_BUTTON}
              </Button>
            </Form>
          );
        }}
      </Formik>
    </div>
  );
};
