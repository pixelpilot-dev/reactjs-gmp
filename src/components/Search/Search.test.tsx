import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import '@testing-library/jest-dom/extend-expect';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { Search } from './Search';

const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as any),
  useNavigate: () => mockedUsedNavigate,
}));

describe('Search component', () => {
  it('should render correctly', () => {
    const { asFragment } = render(<Search />, { wrapper: BrowserRouter });

    expect(asFragment()).toMatchSnapshot();
  });

  it('should navigate after submit', async () => {
    const user = userEvent.setup();
    render(<Search />, { wrapper: BrowserRouter });

    const input = screen.getByPlaceholderText(/What do you want to watch/i);
    expect(input).toBeInTheDocument();
    await user.type(input, 'Alice');

    const button = screen.getByRole('button', { name: /search/i });
    expect(button).toBeInTheDocument();
    await user.click(button);

    await waitFor(() =>
      expect(mockedUsedNavigate).toHaveBeenCalledWith('/search/Alice', { replace: true }),
    );
  });
});
