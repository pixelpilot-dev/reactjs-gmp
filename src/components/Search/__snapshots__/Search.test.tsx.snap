// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component should render correctly 1`] = `
<DocumentFragment>
  <div
    class="search"
  >
    <div
      class="title"
    >
      Find your movie
    </div>
    <form
      action="#"
      class="form"
    >
      <input
        class="input input"
        id="search"
        name="search"
        placeholder="What do you want to watch?"
        type="text"
        value=""
      />
      <button
        class="button default button"
        type="submit"
      >
        Search
      </button>
    </form>
  </div>
</DocumentFragment>
`;
